README.md
# wai-middleware-auth

[![Build Status](https://github.com/fpco/wai-middleware-auth/actions/workflows/tests.yml/badge.svg)](https://github.com/fpco/wai-middleware-auth/actions/workflows/tests.yml)

Middleware that secures WAI application

## Installation

```shell
$ stack install wai-middleware-auth
```
OR
```shell
$ cabal install wai-middleware-auth
```

## wai-auth

Along with middleware this package ships with an executable `wai-auth`, which
can function as a protected file server or a reverse proxy. Right from the box
it supports OAuth2 authentication as well as it's custom implementations for
Google and Github.

Configuration is done using a yaml config file. Here is a sample file that will
configure `wai-auth` to run a file server with Google, GitHub, and GitLab
authentication on `http://localhost:3000`:

```yaml
app_root: "_env:APPROOT:http://localhost:3000"
app_port: 3000
cookie_age: 3600
secret_key: "...+vwscbKR4DyPT"
file_server:
  root_folder: "/path/to/html/files"
  redirect_to_index: true
  add_trailing_slash: true
providers:
  github:
    client_id: "...94cc"
    client_secret: "...166f"
    app_name: "Dev App for wai-middleware-auth"
    email_white_list:
      - "^[a-zA-Z0-9._%+-]+@example.com$"
  google:
    client_id: "...qlj.apps.googleusercontent.com"
    client_secret: "...oxW"
    email_white_list:
      - "^[a-zA-Z0-9._%+-]+@example.com$"
  gitlab:
    client_id: "...9cfc"
    client_secret: "...f0d0"
    app_name: "Dev App for wai-middleware-auth"
    email_white_list:
      - "^[a-zA-Z0-9._%+-]+@example.com$"
```

Above configuration will also block access to users that don't have an email
with `example.com` domain. There is also a `secret_key` field which will be used
to encrypt the session cookie. In order to generate a new random key run this command:

```shell
$ echo $(wai-auth key --base64)
azuCFq0zEBkLSXhQrhliZzZD8Kblo...
```

Make sure you have proper callback/redirect urls registered with
google/github/gitlab apps, eg:
`http://localhost:3000/_auth_middleware/google/complete`.

After configuration file is ready, running application is very easy:

```shell
$ wai-auth --config-file=/path/to/config.yaml
Listening on port 3000
```

### Reverse proxy

To use a reverse proxy instead of a file server, replace `file_server` with
`reverse_proxy`, eg:

```yaml
reverse_proxy:
  host: myapp.example.com
  port: 80
  secure: false
```

### Self-hosted GitLab

The GitLab provider also supports using a self-hosted GitLab instance by
setting the `gitlab_host` field.  In this case you may also want to override
the `provider_info` to change the title, logo, and description.  For example:

```yaml
providers:
  gitlab:
    gitlab_host: gitlab.mycompany.com
    client_id: "...9cfc"
    client_secret: "...f0d0"
    app_name: "Dev App for wai-middleware-auth"
    email_white_list:
      - "^[a-zA-Z0-9._%+-]+@mycompany.com$"
    provider_info:
      title: My Company's GitLab
      logo_url: https://mycompany.com/logo.png
      descr: Use your My Company GitLab account to access this page.
```


CHANGELOG.md
# 0.2.6.0

- Add support for https in reverse_proxy

# 0.2.5.1

- Add support for GHC 9.0.

# 0.2.5.0

- Add `getAuthUserFromVault` for `Servant.Api.Vault` user.

# 0.2.4.0

- Add GitLab provider.

# 0.2.3.1

- Expose `discoverURI` in `Network.Wai.Middleware.Auth.OIDC`
- Fix bug with OAuth2 and OpenID Connect authentication where scopes were
  separated using comma's instead of spaces.

  # 0.2.3.0

- Support `hoauth2-1.11.0`
- Drop support for `jose` versions < 0.8
- Expose `decodeKey`
- OAuth2 provider remove a session when an access token expires. It will use a
  refresh token if one is available to create a new session. If no refresh token
  is available it will redirect the user to re-authenticate.
- Providers can define logic for refreshing a session without user intervention.
- Add an OpenID Connect provider.

  # 0.2.2.0

- Add request logging to executable
- Newer multistage Docker build system

  # 0.2.1.0

- Fix a bug in deserialization of `UserIdentity`

  # 0.2.0.0

- Drop compatiblity with hoauth2 versions <= 1.0.0.
- Add a function for getting the oauth2 token from an authenticated request.
- Modify encoding of oauth2 session cookies. As a consequence existing cookies will be invalid.

  # 0.1.2.1

- Compatibility with hoauth2-1.3.0 - fixed: [#4](https://github.com/fpco/wai-middleware-auth/issues/4)

  # 0.1.2.0

- Implemented compatibility with hoauth2 >= 1.0.0 - fixed: [#3](https://github.com/fpco/wai-middleware-auth/issues/3)

  # 0.1.1.2

- Fixed [wai-middleware-auth-0.1.1.1 does not compile in 32 bit Linux](https://github.com/fpco/wai-middleware-auth/issues/2)

  # 0.1.1.1

- Disallow empty `userIdentity` to produce a successfull login.
- Produces a 404 on `/favicon.ico` page if not logged in: work around for issue
  with Chrome requesting it first and messing up the redirect url.
- Added JQuery to the template, since it's bootstrap's requirement.

  # 0.1.1.0

- Fixed whitelist email regex matching for Github and Google auth.

  # 0.1.0.0

- Initial implementation.


src/Network/Wai/Auth/Config.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}
{-# LANGUAGE TemplateHaskell   #-}
module Network.Wai.Auth.Config
  ( AuthConfig(..)
  , SecretKey(..)
  , Service(..)
  , FileServer(..)
  , ReverseProxy(..)
  , encodeKey
  , decodeKey
  ) where

import           Data.Aeson
import           Data.Aeson.TH          (deriveJSON)
import qualified Data.Text              as T
import           Data.Text.Encoding     (encodeUtf8)
import           Network.Wai.Auth.Tools (decodeKey, encodeKey,
                                         toLowerUnderscore)
import           Web.ClientSession      (Key)

-- | Configuration for a secret key that will be used to encrypt authenticated
-- user as client side cookie.
data SecretKey
  = SecretKeyFile FilePath -- ^ Path to a secret key file in binary form, if it
                           -- is malformed or doesn't exist it will be
                           -- (re)created. If empty "client_session_key.aes"
                           -- name will be used
  | SecretKey Key -- ^ Serialized and base64 encoded form of a secret key. Use
                  -- `encodeKey` to get a proper encoded form.


-- | Configuration for reverse proxy application.
data FileServer = FileServer
    { fsRootFolder       :: FilePath -- ^ Path to a folder containing files
                                     -- that will be served by this app.
    , fsRedirectToIndex  :: Bool -- ^ Redirect to the actual index file, not
                                 -- leaving the URL containing the directory
                                 -- name
    , fsAddTrailingSlash :: Bool -- ^ Add a trailing slash to directory names
    }

-- | Configuration for reverse proxy application.
data ReverseProxy = ReverseProxy
    { rpHost   :: T.Text -- ^ Hostname of the destination webserver
    , rpPort   :: Int -- ^ Port of the destination webserver
    , rpSecure :: Maybe Bool -- ^ Should the request be sent to destination webbserver using https or not (default: false)
    }

-- | Available services.
data Service = ServiceFiles FileServer
             | ServiceProxy ReverseProxy

-- | Configuration for @wai-auth@ executable and any other, that is created using
-- `Network.Wai.Auth.Executable.mkMain`
data AuthConfig = AuthConfig
  { configAppRoot    :: Maybe T.Text -- ^ Root Url of the website, eg:
                                     -- http://example.com or
                                     -- https://example.com It will be used to
                                     -- perform redirects back from external
                                     -- authentication providers.
  , configAppPort    :: Int  -- ^ Port number. Default is 3000
  , configRequireTls :: Bool -- ^ Require requests come in over a secure
                             -- connection (determined via headers). Will
                             -- redirect to HTTPS if non-secure
                             -- dedected. Default is @False@
  , configSkipAuth   :: Bool -- ^ Turn off authentication middleware, useful for
                             -- testing. Default is @False@
  , configCookieAge  :: Int -- ^ Duration of the session in seconds. Default is
                            -- one hour (3600 seconds).
  , configSecretKey  :: SecretKey -- ^ Secret key. Default is "client_session_key.aes"
  , configService    :: Service
  , configProviders  :: Object
  }

$(deriveJSON defaultOptions { fieldLabelModifier = toLowerUnderscore . drop 2} ''FileServer)

$(deriveJSON defaultOptions { fieldLabelModifier = toLowerUnderscore . drop 2} ''ReverseProxy)

instance FromJSON AuthConfig where
  parseJSON =
    withObject "Auth Config Object" $ \obj -> do
      configAppRoot <- obj .:? "app_root"
      configAppPort <- obj .:? "app_port" .!= 3000
      configRequireTls <- (obj .:? "require_tls" .!= False)
      configSkipAuth <- obj .:? "skip_auth" .!= False
      configCookieAge <- obj .:? "cookie_age" .!= 3600
      mSecretKeyB64T <- obj .:? "secret_key"
      configSecretKey <-
        case mSecretKeyB64T of
          Just secretKeyB64T ->
            either fail (return . SecretKey) $ decodeKey (encodeUtf8 secretKeyB64T)
          Nothing -> SecretKeyFile <$> (obj .:? "secret_key_file" .!= "")
      mFileServer <- obj .:? "file_server"
      mReverseProxy <- obj .:? "reverse_proxy"
      let sErrMsg =
            "Either 'file_server' or 'reverse_proxy' is required, but not both."
      configService <-
        case (mFileServer, mReverseProxy) of
          (Just fileServer, Nothing) -> ServiceFiles <$> parseJSON fileServer
          (Nothing, Just reverseProxy) ->
            ServiceProxy <$> parseJSON reverseProxy
          (Just _, Just _) -> fail $ "Too many services. " ++ sErrMsg
          (Nothing, Nothing) -> fail $ "No service is supplied. " ++ sErrMsg
      configProviders <- obj .: "providers"
      return AuthConfig {..}
